src/streams.rs

```diff
@@ -0,0 +1,669 @@
+    /// last time this message was delivered to this consumer.
+    pub last_delivered_ms: usize,
+    /// The number of times this message was delivered.
+    pub times_delivered: usize,
+}
+
+/// Represents a stream `key` and its `id`'s parsed from `xread` methods.
+#[derive(Default, Debug, Clone)]
+pub struct StreamKey {
+    /// The stream `key`.
+    pub key: String,
+    /// The parsed stream `id`'s.
+    pub ids: Vec<StreamId>,
+}
+
+impl StreamKey {
+    /// Return only the stream `id`'s without their data.
+    pub fn just_ids(&self) -> Vec<&String> {
+        self.ids.iter().map(|msg| &msg.id).collect::<Vec<&String>>()
+    }
+}
+
+/// Represents a stream `id` and its field/values as a `HashMap`
+#[derive(Default, Debug, Clone)]
+pub struct StreamId {
+    /// The stream `id` (entry ID) of this particular message.
+    pub id: String,
+    /// All fields in this message, associated with their respective values.
+    pub map: HashMap<String, Value>,
+}
+
+impl StreamId {
+    /// Converts a `Value::Bulk` into a `StreamId`.
+    pub fn from_bulk_value(v: &Value) -> RedisResult<Self> {
+        let mut stream_id = StreamId::default();
+        if let Value::Bulk(ref values) = *v {
+            if let Some(v) = values.get(0) {
+                stream_id.id = from_redis_value(&v)?;
+            }
+            if let Some(v) = values.get(1) {
+                stream_id.map = from_redis_value(&v)?;
+            }
+        }
+
+        Ok(stream_id)
+    }
+
+    /// Fetches value of a given field and converts it to the specified
+    /// type.
+    pub fn get<T: FromRedisValue>(&self, key: &str) -> Option<T> {
+        match self.find(&key) {
+            Some(ref x) => from_redis_value(*x).ok(),
+            None => None,
+        }
+    }
+
+    /// Fetches the value of a given field without performing any
+    /// type conversion.
+    pub fn find(&self, key: &&str) -> Option<&Value> {
+        self.map.get(*key)
+    }
+
+    /// Does the message contain a particular field?
+    pub fn contains_key(&self, key: &&str) -> bool {
+        self.find(key).is_some()
+    }
+
+    /// Returns how many field/value pairs exist in this message.
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Returns true if there are no field/value pairs in this message.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+type SRRows = Vec<HashMap<String, Vec<HashMap<String, HashMap<String, Value>>>>>;
+impl FromRedisValue for StreamReadReply {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        let rows: SRRows = from_redis_value(v)?;
+        let mut reply = StreamReadReply::default();
+        for row in rows {
+            for (key, entry) in row.iter() {
+                let mut k = StreamKey::default();
+                k.key = key.to_owned();
+                for id_row in entry[..].iter() {
+                    for (id, map) in id_row.iter() {
+                        k.ids.push(StreamId {
```

Comment:
@grippy: I altered the structure of these loops based on @Marwes's comment in https://github.com/mitsuhiko/redis-rs/pull/319#discussion_r434025119.  Let me know if we're wrong about this, but I think we want all the `StreamId`s added to each `StreamKey`.